File: 0726-number-of-atoms/0726-number-of-atoms.cs
public class Solution {
    private int i;
    public string CountOfAtoms(string formula) {
        StringBuilder ans = new StringBuilder();
        i = 0;
        var counts = CountOfAtoms(formula.ToCharArray());
        foreach (var name in counts.Keys) {
            ans.Append(name);
            int count = counts[name];
            if (count > 1) ans.Append(count);
        }
        return ans.ToString();
    }
    
    private SortedDictionary<string, int> CountOfAtoms(char[] f) {
        var ans = new SortedDictionary<string, int>();
        while (i < f.Length) {
            if (f[i] == '(') {
                i++;
                var tmp = CountOfAtoms(f);
                int count = GetCount(f);
                foreach (var entry in tmp) {
                    if (ans.ContainsKey(entry.Key)) {
                        ans[entry.Key] += entry.Value * count;
                    } else {
                        ans[entry.Key] = entry.Value * count;
                    }
                }
            } else if (f[i] == ')') {
                i++;
                return ans;
            } else {
                string name = GetName(f);
                if (ans.ContainsKey(name)) {
                    ans[name] += GetCount(f);
                } else {
                    ans[name] = GetCount(f);
                }
            }
        }
        return ans;
    }

    private string GetName(char[] f) {
        StringBuilder name = new StringBuilder();
        name.Append(f[i++]);
        while (i < f.Length && char.IsLower(f[i])) {
            name.Append(f[i++]);
        }
        return name.ToString();
    }

    private int GetCount(char[] f) {
        int count = 0;
        while (i < f.Length && char.IsDigit(f[i])) {
            count = count * 10 + (f[i] - '0');
            i++;
        }
        return count == 0 ? 1 : count;
    }
}

File: 0726-number-of-atoms/0726-number-of-atoms.cpp
#include<bits/stdc++.h>
class Solution {
public:
    string countOfAtoms(string formula) {
        stack<unordered_map<string, int>>st;
        st.push(unordered_map<string, int>());
        
        int index = 0;
        
        while(index < formula.length())
        {
            if(formula[index] == '(')
            {
                st.push(unordered_map<string, int>());
                index++;
            }
            
            else if(formula[index] == ')')
            {
                unordered_map<string, int> currMap = st.top();
                st.pop();
                index++;
                string multiplier;
                while(index < formula.length() && isdigit(formula[index]))
                {
                    multiplier += formula[index];
                    index++;
                }
                
                if(!multiplier.empty())
                {
                    int mult = stoi(multiplier);
                    for(auto& [atom, count] : currMap)
                    {
                        currMap[atom] = count*mult;
                    }
                }
                
                for(auto & [atom, count]: currMap)
                {
                    st.top()[atom] += count;
                }
            }
            
            else
            {
                string currAtom;
                currAtom += formula[index];
                index++;
                
                while(index < formula.length() && islower(formula[index]))
                {
                    currAtom += formula[index];
                    index++;
                }
                
                string currCount;
                while(index < formula.length() && isdigit(formula[index]))
                {
                    currCount += formula[index];
                    index++;
                }
                
                int count = currCount.empty() ? 1 : stoi(currCount);
                st.top()[currAtom] += count;
            }
        }
        
        map<string, int> finalMap(st.top().begin(), st.top().end());
        
        string ans;
        for(auto& [atom, count] : finalMap)
        {
            ans += atom;
            if(count > 1)
            {
                ans += to_string(count);
            }
        }
        
        return ans;
    } 
};